wallet: when a block is disconnected, update transactions that are no longer conflicted

ishaanam opened the pull request:
This implements a fix for #7315. Previously when a block was disconnected any transactions that were conflicting with transactions mined in that block were not updated to be marked as inactive. The fix implemented here is described on the [Bitcoin DevWiki](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking#idea-refresh-conflicted). A test which tested the previous behavior has also been updated.

Second attempt at #17543

Commit:
Add wallets_conflicts

Test the case of a tx being conflicted by multiple
txs with different depths. The conflicted tx is also spent by
a child tx for which confirmation status is tied to the parent's.
After a reorg of conflicting txs, the conflicted status should be
undone properly.

Co-authored-by: furszy <mfurszy@protonmail.com>

ryanofsky commented:
Concept ACK on refreshing state of conflicted transactions in the wallet when a block is disconnected. I think there is a potential problem with the implementation though because it isn't recursively marking conflicted transactions as unconfirmed, only marking the top level transactions without recursing.

I think a good approach could be to steal from the previous implementation of this idea in commits b8b4592d985fa36dd0dddfd0e2bc6daf8df6e79c and 3c8bd6814ca88715d3d56994b10dc2c5f3e38e2d from #17543, which did recurse. That approach was also nice because it reused the existing MarkConflicted function in order to avoid code duplication. Commit ec5a89a26fa17e68b3e5928730e7d7c3e0370729 from that PR also added some more test coverage which could be used here.

ishaanam committed:
wallet, tests: mark unconflicted txs as inactive

In `blockDisconnected`, for each transaction in the block, look
for any wallet transactions spending the same inputs. If any of
these transactions were marked conflicted, they are now marked as
inactive.

Co-authored-by: ariard <antoine.riard@gmail.com>

ishaanam commented:
@ryanofsky 
> I think there is a potential problem with the implementation though because it isn't recursively marking conflicted transactions as unconfirmed, only marking the top level transactions without recursing.

Yes, I had forgotten about this, thanks for pointing it out! 
> I think a good approach could be to steal from the previous implementation of this idea in commits https://github.com/bitcoin/bitcoin/commit/b8b4592d985fa36dd0dddfd0e2bc6daf8df6e79c and https://github.com/bitcoin/bitcoin/commit/3c8bd6814ca88715d3d56994b10dc2c5f3e38e2d from https://github.com/bitcoin/bitcoin/pull/17543, which did recurse.

I have used ideas from that PR so that the transaction state updating will recurse.
> Commit https://github.com/bitcoin/bitcoin/commit/ec5a89a26fa17e68b3e5928730e7d7c3e0370729 from that PR also added some more test coverage which could be used here.

I have cherry-picked that commit to this branch with a few adjustments.

glozow commented on test/functional/wallet_conflicts.py:
```diff
@@ -0,0 +1,113 @@
+        # Conflicted tx should have confirmations set to the confirmations of the most conflicting tx
+        assert_equal(conflicted["confirmations"], -conflicting["confirmations"])
+        # Child should inherit conflicted state from parent
+        assert_equal(child_conflicted["confirmations"], -conflicting["confirmations"])
```
We should also test who "wins" in this conflict contest and what the values are, i.e. conflicting A has 8 confirmations and conflicting B has 4 confirmations.

ishaanam replied:
I've added a check for this.

glozow commented on test/functional/wallet_conflicts.py:
```diff
@@ -0,0 +1,113 @@
+        self.log.info("Create a transaction with conflicted transactions spending outpoints A & B")
+        nA = next(tx_out["vout"] for tx_out in self.nodes[0].gettransaction(txid_conflict_from_1)["details"] if tx_out["amount"] == Decimal("10"))
+        nB = next(tx_out["vout"] for tx_out in self.nodes[0].gettransaction(txid_conflict_from_2)["details"] if tx_out["amount"] == Decimal("10"))
+        inputs_conflicted_tx = []
+        inputs_conflicted_tx.append({"txid": txid_conflict_from_1, "vout": nA})
+        inputs_conflicted_tx.append({"txid": txid_conflict_from_2, "vout": nB})
```
nit in 77a2e522d845eda270b2c03fd7e0a16bc4c00fbb, I think this is clearer (same with the other arrays of inputs and outputs)

```suggestion
        inputs_conflicted_tx = [{"txid": txid_conflict_from_1, "vout": nA}, {"txid": txid_conflict_from_2, "vout": nB}]
```

ishaanam replied:
Done

glozow commented on test/functional/wallet_conflicts.py:
```diff
@@ -0,0 +1,113 @@
+        outputs_conflicting_tx_B = {}
+        outputs_conflicting_tx_B[self.nodes[0].getnewaddress()] = Decimal("9.99998")
+
+        conflicted = self.nodes[0].signrawtransactionwithwallet(self.nodes[0].createrawtransaction(inputs_conflicted_tx, outputs_conflicted_tx))
+        conflicting_tx_A = self.nodes[0].signrawtransactionwithwallet(self.nodes[0].createrawtransaction(inputs_conflicting_tx_A, outputs_conflicting_tx_A))
+        conflicting_tx_B = self.nodes[0].signrawtransactionwithwallet(self.nodes[0].createrawtransaction(inputs_conflicting_tx_B, outputs_conflicting_tx_B))
```
Sorry to bikeshed, but I think the naming is a bit hard to follow ("conflicted" vs "conflicting" don't immediately make sense to me, especially since both sets of transactions get conflicted at some point during the test). Might I suggest names that indicate which input(s) they spend, which node broadcasts them, and what "generation" they are? For example

`tx_AB_0_parent` instead of `conflicted`
`tx_A_1` instead of `conflicting_tx_A`
`tx_B_1` instead of `conflicting_tx_B`

`C` instead of `nA` for the 19.99998BTC output, because `nA` is already the name of the 10BTC output from one of the funding transactions.
`tx_C_0_child` instead of `child_conflicted`

ishaanam replied:
I agree, the naming here was a bit confusing. It should be fixed now.

glozow commented on test/functional/wallet_conflicts.py:
```diff
@@ -0,0 +1,113 @@
+
+        # Connect node0 and node2 and wait reorg
+        self.connect_nodes(0, 2)
+        self.sync_blocks([self.nodes[0], self.nodes[2]])
```
```suggestion
        self.sync_blocks()
```

ishaanam replied:
Done both here and one other place.

glozow reviewed (commented):
Concept ACK

There is a missing dash in "Co-authored-by:" in the commit messages

ishaanam commented on test/functional/wallet_conflicts.py:
```diff
@@ -0,0 +1,113 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""Test conflicts tracking with multilple txn conflicting a conflicted tx."""
```
I've updated it to this, this is much more clear.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1465,8 +1465,32 @@ void CWallet::blockDisconnected(const interfaces::BlockInfo& block)
     m_last_block_processed_height = block.height - 1;
     m_last_block_processed = *Assert(block.prev_hash);
+
+    const int& disconnect_height = block.height;
```
Fixed

ishaanam commented:
Thanks for the review @glozow, I have addressed your comments and also made a modification to `RecursiveUpdateTxState` which will allow it to be more extensible.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1336,13 +1336,29 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
+
+}
+
+void CWallet::RecursiveUpdateTxState(const uint256& tx_hash, const GetUpdatedStateFn& get_updated_state) {
```
In c3fe92175b6ac709e4fdfd4766253c75d7ffd184 "wallet: introduce generic recursive tx state updating function"

This could also be used by `AbandonTransaction`.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1351,11 +1367,7 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
+        if (get_updated_state(wtx)) {
             wtx.MarkDirty();
             batch.WriteTx(wtx);
             // Iterate over all its outputs, and mark transactions in the wallet that spend them conflicted too
```
In c3fe92175b6ac709e4fdfd4766253c75d7ffd184 "wallet: introduce generic recursive tx state updating function"

The comments should be updated to avoid mentioning specific transaction states.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1336,13 +1313,29 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
+
+}
+
+void CWallet::RecursiveUpdateTxState(const uint256& tx_hash, const GetUpdatedStateFn& get_updated_state) {
```
Yes, I've changed `GetUpdatedStateFn` to `TryUpdatingStateFn`, and `get_updated_state` to `try_updating_state`.

furszy reviewed (commented):
Haven't finished checking the latest changes but, if you like it, https://github.com/furszy/bitcoin-core/commit/8bb14c1eaf09b0bca9efa78e1a7a7a00c61c0a29 is all yours. During the review, the functional test wasn't clear enough for me so I spent a bit of time making it more friendly.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1336,13 +1313,27 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
+}
+
+void CWallet::RecursiveUpdateTxState(const uint256& tx_hash, const TryUpdatingStateFn& try_updating_state) {
+    LOCK(cs_wallet);
```
In f998cdbc:
This lock isn't needed. All callers have `cs_wallet` acquired. Could add the lock required annotation instead.

ishaanam replied:
Done

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1282,44 +1277,26 @@ bool CWallet::AbandonTransaction(const uint256& hashTx)
             wtx.m_state = TxStateInactive{/*abandoned=*/true};
-            wtx.MarkDirty();
-            batch.WriteTx(wtx);
             NotifyTransactionChanged(wtx.GetHash(), CT_UPDATED);
```
In f998cdbc:

Would be good to move this `NotifyTransactionChanged` to `RecursiveUpdateTxState` too. So `MarkConflicted` and the `BlockDisconnected` txes state changes are notified to the upper layers.

ishaanam replied:
I think that this is a good idea, but would out of the scope of this PR and would warrant more discussion, since then it would be notifying for tx state updates that it previously did not.

ryanofsky replied:
In commit "wallet: introduce generic recursive tx state updating function" (f998cdbc689cc65ff0aca98ce63758c59786c1db)

> I think that this is a good idea, but would out of the scope of this PR and would warrant more discussion, since then it would be notifying for tx state updates that it previously did not.

Agree we should avoid changing behavior, but I think the current commit might also be changing behavior by notifying before calling MarkDirty and WriteTx instead of after.

I think the best thing to do would be to change TryUpdatingStateFn to return an enum instead of a bool like:

```c++
enum class TxUpdate { UNCHANGED, CHANGED, NOTIFY_CHANGED };
using TryUpdatingStateFn = std::function<TxUpdate(CWalletTx& wtx)>;
```

Returning CHANGED/UNCHANGED should make the code clearer anyway because it's less obvious what true and false mean.

ishaanam replied:
I have implemented this suggestion, I agree that it is much clearer now.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1443,8 +1443,36 @@ void CWallet::blockDisconnected(const interfaces::BlockInfo& block)
+
+                if (!wtx.isConflicted()) continue;
+
+                auto try_updating_state = [&](CWalletTx& wtx) {
```
> The lambda function's wtx shadows the function's wtx.

I have renamed the lambda function's `wtx`.
> Could just provide disconnect_height.

By this do you mean providing the `disconnect_height` of the block to `RecursiveUpdateTxState` or providing the wallet tx's `conflicting_block_height` to `try_updating_state` instead of the whole wallet tx by reference?

furszy replied:
Yeah, nothing biggie, just:
```c++
auto try_updating_state = [disconnect_height](CWalletTx& tx) {
```
so only the required elements are provided to the lambda.

ishaanam commented:
Thanks for the review @achow101 and @furszy!
> Haven't finished checking the latest changes but, if you like it, https://github.com/furszy/bitcoin-core/commit/8bb14c1eaf09b0bca9efa78e1a7a7a00c61c0a29 is all yours. During the review, the functional test wasn't clear enough for me so I spent a bit of time making it more friendly.

Thanks, I'll use these improvements here once I review them.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1336,13 +1313,27 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
     if (conflictconfirms >= 0)
         return;
 
-    // Do not flush the wallet here for performance reasons
```
In commit "wallet: introduce generic recursive tx state updating function" (f998cdbc689cc65ff0aca98ce63758c59786c1db)

This also seems like a useful comment. Would keep instead of deleting

ishaanam replied:
I have added it back.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1336,13 +1313,27 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
+        return false;
+    };
+
+    RecursiveUpdateTxState(hashTx, try_updating_state);
```
In commit "wallet: introduce generic recursive tx state updating function" (f998cdbc689cc65ff0aca98ce63758c59786c1db)

This code pretty opaque, would suggest copying the previous comment here "Iterate over all its outputs, and mark transactions in the wallet that spend them conflicted too" or something equivalent.

ishaanam replied:
Done

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1336,13 +1313,27 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
-    // Do not flush the wallet here for performance reasons
+    auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+        if (conflictconfirms < GetTxDepthInMainChain(wtx)) {
+            wtx.m_state = TxStateConflicted{hashBlock, conflicting_height};
```
In commit "wallet: introduce generic recursive tx state updating function" (f998cdbc689cc65ff0aca98ce63758c59786c1db)

Would suggest keeping the previous comment here instead of deleting it. It's not obvious what the check means or what's happening in the block:

```c++
// Block is 'more conflicted' than current confirm; update.
// Mark transaction as conflicted with this block.
```

ishaanam replied:
Done

ishaanam committed:
wallet: introduce generic recursive tx state updating function

This commit also changed `MarkConflicted` and `AbandonTransaction`
to use this new function

Co-authored-by: ariard <antoine.riard@gmail.com>

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1282,44 +1277,26 @@ bool CWallet::AbandonTransaction(const uint256& hashTx)
         int currentconfirm = GetTxDepthInMainChain(wtx);
-        // If the orig tx was not in block, none of its spends can be
         assert(currentconfirm <= 0);
-        // if (currentconfirm < 0) {Tx and spends are already conflicted, no need to abandon}
```
I implemented your next suggestion about simplifying the code in the lambda, so the comments are a bit different now but I did add them back.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1282,44 +1277,26 @@ bool CWallet::AbandonTransaction(const uint256& hashTx)
-        assert(it != mapWallet.end());
-        CWalletTx& wtx = it->second;
+    auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
         int currentconfirm = GetTxDepthInMainChain(wtx);
```
This is simpler and it is a small change so I've implemented it in this commit.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1282,44 +1277,25 @@ bool CWallet::AbandonTransaction(const uint256& hashTx)
-        if (currentconfirm == 0 && !wtx.isAbandoned()) {
-            // If the orig tx was not in block/mempool, none of its spends can be in mempool
-            assert(!wtx.InMempool());
+    auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
```
Can remove the context ref.
```suggestion
    auto try_updating_state = [](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
```

ishaanam replied:
Done

ishaanam commented:
Rebased to fix failing CI.

furszy commented on test/functional/wallet_abandonconflict.py:
```diff
@@ -226,20 +226,16 @@ def run_test(self):
-        # There is currently a minor bug around this and so this test doesn't work.  See Issue #7315
-        # Invalidate the block with the double spend and B's 10 BTC output should no longer be available
-        # Don't think C's should either
+        # Invalidate the block with the double spend. B & C's 10 BTC outputs should no longer be available
```
Let me see if I'm following this:

`Tx_AB1` spends B, and `Tx_ABC2` (the child of `Tx_AB1`) spends C. So, when the double spend tx gets disconnected, `Tx_AB1` state changes to `inactive`, marking output B and C as spent again.

Which ends up with the txes stuck now right?, and the user need to manually re-submit them.

furszy replied:
good. Not sure if there is any but would be nice to have coverage for the automatic re-submission. Would just need to mock the node time 24hs in the future and see if they get re-submitted.

ishaanam replied:
There are some tests for that here: https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_resendwallettransactions.py. Or did you mean that a test for that should be added for this PR specifically?

furszy replied:
yeah, great if that is already covered.
(sorry for the delayed answer)

furszy reviewed (approved):
ACK 6f1d3948

ishaanam commented:
I've fixed a bug which resulted in the conflicted tx not being marked as inactive if the conflicting tx was not in `mapTxSpends`.

rajarshimaitra reviewed (commented):
tACK 89df7987c2f1eea42454c2b0efc31a924fbfd3a8.

Changes look very clean. We reviewed it in out weekly club and went over the new code addition. 

Ran the tests on the master branch and saw it failing `assert_equal(conflicted["confirmations"], 0)`. In the master conflicting transaction's state doesn't update after reorg.

achow101 commented:
ACK 89df7987c2f1eea42454c2b0efc31a924fbfd3a8

furszy commented:
> I've fixed a bug which resulted in the conflicted tx not being marked as inactive if the conflicting tx was not in mapTxSpends.

Could you expand this?

Not sure if I follow the [latest push](https://github.com/bitcoin/bitcoin/compare/6f1d394896fcc7e2687b766418f9cd4e55b613b5..89df7987c2f1eea42454c2b0efc31a924fbfd3a8), where only the inequality was changed.
If `mapTxSpends.count(outpoint) == 1`, the outpoint was spent only once, so the wallet shouldn't contain any conflicting tx.

glozow commented:
> I've fixed a bug which resulted in the conflicted tx not being marked as inactive if the conflicting tx was not in mapTxSpends.

> If mapTxSpends.count(outpoint) == 1, the outpoint was spent only once, so the wallet shouldn't contain any conflicting tx.

I interpreted the bug as = we wouldn't mark a conflicted tx inactive if the block transaction was not in `mapTxSpends` (which would mean count == 1). Normally we'd expect them both to be there (count > 1).

glozow reviewed (commented):
ACK 89df7987c2f1eea42454c2b0efc31a924fbfd3a8

ishaanam commented:
>  If mapTxSpends.count(outpoint) == 1, the outpoint was spent only once, so the wallet shouldn't contain any conflicting tx.

This is not necessarily always true, because even though the conflicted wallet transaction must always have its spends added to `mapTxSpends`, this is not the case for all conflicting transactions. 

For example, let's say that Alice has a utxo. She creates a transaction, tx1, spending this utxo to Bob. `AddToSpends` will get called on Bob's node for this transaction, because Bob is the recipient. Alice then creates a conflicting transaction to tx1, called tx2, which sends the funds back to herself. However, Bob's node does not call `AddToSpends` on tx2, because Bob is not the recipient. If tx2 gets mined, but then that block is disconnected from our node, Bob's `mapTxSpends` will only show a single spend for Alice's utxo in `mapTxSpends`, even though there are technically two.

furszy commented:
Ok nice, great. Thanks for the details.
We should code a test exercising exactly that behavior.

furszy reviewed (approved):
Tested ACK 89df7987

> For example, let's say that Alice has a utxo. She creates a transaction, tx1, spending this utxo to Bob. AddToSpends will get called on Bob's node for this transaction, because Bob is the recipient. Alice then creates a conflicting transaction to tx1, called tx2, which sends the funds back to herself. However, Bob's node does not call AddToSpends on tx2, because Bob is not the recipient. If tx2 gets mined, but then that block is disconnected from our node, Bob's mapTxSpends will only show a single spend for Alice's utxo in mapTxSpends, even though there are technically two.

Made a quick unit test and all good. Good finding.
Would be good to have a real functional test in the future.

ishaanam commented:
> Would be good to have a real functional test in the future.

#27307 tests this behavior, which is how I found this bug.
